prow.sh

```diff
@@ -0,0 +1,667 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# This script runs inside a Prow job. It runs unit tests ("make test")
+# and E2E testing. This E2E testing covers different scenarios (see
+# https://github.com/kubernetes/enhancements/pull/807):
+# - running the stable hostpath example against a Kubernetes release
+# - running the canary hostpath example against a Kubernetes release
+# - building the component in the current repo and running the
+#   stable hostpath example with that one component replaced against
+#   a Kubernetes release
+#
+# The intended usage of this script is that individual repos import
+# csi-release-tools, then link their top-level prow.sh to this or
+# include it in that file. When including it, several of the variables
+# can be overridden in the top-level prow.sh to customize the script
+# for the repo.
+#
+# The expected environment is:
+# - $GOPATH/src/<import path> for the repository that is to be tested,
+#   with PR branch merged (when testing a PR)
+# - running on linux-amd64
+# - bazel installed (when testing against Kubernetes master), must be recent
+#   enough for Kubernetes master
+# - kind (https://github.com/kubernetes-sigs/kind) installed
+# - optional: Go already installed
+
+# Sets the default value for a variable if not set already and logs the value.
+# Any variable set this way is usually something that a repo's .prow.sh
+# or the job can set.
+configvar () {
+    # Ignore: Word is of the form "A"B"C" (B indicated). Did you mean "ABC" or "A\"B\"C"?
+    # shellcheck disable=SC2140
+    eval : \$\{"$1":="\$2"\}
+    eval echo "\$3:" "$1=\${$1}"
+}
+
+# Go version to use. If the pre-installed Go is missing or a different
+# version, this version here will get installed.
+configvar CSI_PROW_GO_VERSION 1.11.4 "Go version"
+
+# kind version to use. If the pre-installed version is different,
+# the desired version is downloaded from https://github.com/kubernetes-sigs/kind/releases/download/
+# (if available), otherwise it is built from source.
+configvar CSI_PROW_KIND_VERSION master "kind"
+
+# ginkgo test runner version to use. If the pre-installed version is
+# different, the desired version is built from source.
+configvar CSI_PROW_GINKGO_VERSION v1.7.0 "Ginkgo"
+
+# Ginkgo runs the E2E test in parallel. The default is based on the number
+# of CPUs, but typically this can be set to something higher in the job.
+configvar CSI_PROW_GINKO_PARALLEL "-p" "Ginko parallelism parameter(s)"
```

Comment:
Those can be part of the alpha test run, which runs sequentially.

I had considered having parameters for parallel, serial, and alpha, but then decided against it because the only difference between serial and alpha would have been the naming of the parameters.

Reply:
I think we should have 4 jobs:
- parallel 
- serial
- parallel alpha
- serial alpha

It will make it easier to rule out if a test failure is due to an alpha feature or not.

Reply:
Okay, I'll add the ability to control which of these run.

But for now let's run them all inside a single job until a) Prow can have such a high number of jobs (right now I think it is still limited by the ConfigMap size) and b) we really know which of these are needed.

For example, there are no serial alpha jobs.

> We have two reconstruction tests (one of which is skipped for hostpath but actually it only needs to be skipped for in-tree and not csi): https://k8s-testgrid.appspot.com/sig-storage-kubernetes#gce-serial&include-filter-by-regex=csi&include-filter-by-regex=CSI.*Volumes

No test is explicitly marked up as "Serial", and the "Disruptive" test cannot be run inside a kind cluster ("No external address for pod pod-subpath-test-csi-hostpath-dynamicpv-4lsw on node csi-prow-control-plane") - or least that is my initial impression. Need to check this error in more detail.